[UNIT] llama-index framework - release 4/3/25 (#22)
huggingface/agents-course: units/en/unit2/llama-index/workflows.mdx

```python
from llama_index.core.workflow import Context, StartEvent, StopEvent

@step
async def query(self, ctx: Context, ev: StartEvent) -> StopEvent:
    # retrieve from context
    query = await ctx.get("query")

    # do something with context and event
    val = ...

    # store in context
    await ctx.set("key", val)

    return StopEvent(result=result)
```

Great! Now you know how to create basic workflows in LlamaIndex!

<Tip>There are some more complex nuances to workflows, which you can learn about in <a href="https://docs.llamaindex.ai/en/stable/understanding/workflows/">the LlamaIndex documentation</a>.</Tip>

However, there is another way to create workflows, which relies on the `AgentWorkflow` class. Let's take a look at how we can use this to create a multi-agent workflow.

## Automating workflows with Multi-Agent Workflows

Instead of manual workflow creation, we can use the **`AgentWorkflow` class to create a multi-agent workflow**.
The `AgentWorkflow` uses Workflow Agents to allow you to create a system of one or more agents that can collaborate and hand off tasks to each other based on their specialized capabilities.

> This enables building complex agent systems where different agents handle different aspects of a task.

Reviewer: ```suggestion

This enables building complex agent systems where different agents handle different aspects of a task.
```

Instead of importing classes from `llama_index.core.agent`, we will import the agent classes from `llama_index.core.agent.workflow`.
One agent must be designated as the root agent in the `AgentWorkflow` constructor.

> When a user message comes in, it is first routed to the root agent. Each agent can then:

Reviewer: ```suggestion
When a user message comes in, it is first routed to the root agent. 

Each agent can then:
```

- Hand off to another agent better suited for the task
- Return a response to the user

Let's see how to create a multi-agent workflow.

```python
from llama_index.core.agent.workflow import AgentWorkflow, FunctionAgent, ReActAgent

query_engine_agent_tool = # as defined in the previous section
# Define the agents
multiply_agent = FunctionAgent(
    fn=lambda x, y: x * y,
    name="multiply",
    description="Multiplies two integers",
)
retriever_agent = ReActAgent(
    llm=llm,
    tools=[query_engine_agent_tool],
)
# Create the workflow
workflow = AgentWorkflow(
    agents=[multiply_agent, retriever_agent], root_agent="multiply"
)

# Run the system
response = await workflow.run(user_msg="Can you add 5 and 3?")
```

Before starting the workflow, we can provide an initial state dict that will be available to all agents.
The state is stored in the state key of the workflow context. It will be injected into the state_prompt which augments each new user message.

```python
workflow = AgentWorkflow(
    agents=[...],
    root_agent="root_agent",
    initial_state={"counter": 0},
    state_prompt="Current state: {state}. User message: {msg}",
)
```

> Congratulations! You have now mastered the basics of Agents in LlamaIndex! 🎉

Reviewer 2: This is kind of already covered in the previous edit I made. It might be more fun to cover a query engine from scratch here maybe?

https://docs.llamaindex.ai/en/stable/examples/workflow/rag/

Or an agent from scratch?

https://docs.llamaindex.ai/en/stable/examples/workflow/function_calling_agent/

Reviewer 2: Or neither, up to you!

However, there is another way to create workflows, which relies on the `AgentWorkflow` class. Let's take a look at how we can use this to create a multi-agent workflow.

## Automating workflows with Multi-Agent Workflows

Instead of manual workflow creation, we can use the **`AgentWorkflow` class to create a multi-agent workflow**.
The `AgentWorkflow` uses Workflow Agents to allow you to create a system of one or more agents that can collaborate and hand off tasks to each other based on their specialized capabilities.
This enables building complex agent systems where different agents handle different aspects of a task.
Instead of importing classes from `llama_index.core.agent`, we will import the agent classes from `llama_index.core.agent.workflow`.
One agent must be designated as the root agent in the `AgentWorkflow` constructor.
When a user message comes in, it is first routed to the root agent.

Each agent can then:

- Handle the request directly using their tools
- Handoff to another agent better suited for the task
- Return a response to the user

Let's see how to create a multi-agent workflow.

> ```python
> from llama_index.core.agent.workflow import AgentWorkflow, FunctionAgent, ReActAgent
> from llama_index.llms.huggingface_api import HuggingFaceInferenceAPI
> 
> multiply_agent = FunctionAgent(
>     fn=lambda x, y: x * y,
>     name="multiply_agent",
>     description="Multiplies two integers",
> )
> addition_agent = FunctionAgent(
>     fn=lambda x, y: x + y,
>     name="addition_agent",
>     description="Adds two integers",
> )
> # Create the workflow
> workflow = AgentWorkflow(
>     agents=[multiply_agent, addition_agent],
>     root_agent="multiply_agent"
> )
> 
> # Run the system
> response = await workflow.run(user_msg="Can you add 5 and 3?")
> ```

Reviewer 2: Just noticed this is incorrect. Let's clean this up a touch

````suggestion
```python
from llama_index.core.agent.workflow import AgentWorkflow, ReActAgent
from llama_index.llms.huggingface_api import HuggingFaceInferenceAPI

# Define some tools
def add(a: int, b: int) -> int:
    """Add two numbers."""
    return a + b

def multiply(a: int, b: int) -> int:
    """Multiply two numbers."""
    return a * b

llm = HuggingFaceInferenceAPI(model_name="Qwen/Qwen2.5-Coder-32B-Instruct")

# we can pass functions directly without FunctionTool -- the fn/docstring are parsed for the name/description
multiply_agent = ReActAgent(
    name="multiply_agent",
    description="Is able to multiply two integers",
    system_prompt="A helpful assistant that can use a tool to multiply numbers.",
    tools=[multiply], 
    llm=llm,
)

addition_agent = ReActAgent(
    name="add_agent",
    description="Is able to add two integers",
    system_prompt="A helpful assistant that can use a tool to add numbers."
    tools=[add], 
    llm=llm,
)

# Create the workflow
workflow = AgentWorkflow(
    agents=[multiply_agent, addition_agent],
    root_agent="multiply_agent"
)

# Run the system
response = await workflow.run(user_msg="Can you add 5 and 3?")
```
````

> Before starting the workflow, we can provide an initial state dict that will be available to all agents.
> The state is stored in the state key of the workflow context. It will be injected into the state_prompt which augments each new user message.
> 
> ```python
> workflow = AgentWorkflow(
>     agents=[...],
>     root_agent="root_agent",
>     initial_state={"counter": 0},
>     state_prompt="Current state: {state}. User message: {msg}",
> )
> ```

Reviewer 2: Just going to suggest to expand this slightly with a solid example

````suggestion
Agent tools can also modify the workflow state we mentioned earlier. Before starting the workflow, we can provide an initial state dict that will be available to all agents.
The state is stored in the state key of the workflow context. It will be injected into the state_prompt which augments each new user message.

Let's inject a counter to count function calls by modifying the previous example:

```python
from llama_index.core.workflow import Context

# Define some tools
async def add(ctx: Context, a: int, b: int) -> int:
    """Add two numbers."""
    # update our count
    cur_state = await ctx.get("state")
    cur_state["num_fn_calls"] += 1
    await ctx.set("state", cur_state)
    
    return a + b
    
async def multiply(ctx: Context, a: int, b: int) -> int:
    """Multiply two numbers."""
    # update our count
    cur_state = await ctx.get("state")
    cur_state["num_fn_calls"] += 1
    await ctx.set("state", cur_state)
    
    return a * b
 
...

workflow = AgentWorkflow(
    agents=[multiply_agent, addition_agent],
    root_agent="multiply_agent"
    initial_state={"num_fn_calls": 0},
    state_prompt="Current state: {state}. User message: {msg}",
)

# run the workflow with context
ctx = Context(workflow)
response = await workflow.run(user_msg="Can you add 5 and 3?", ctx=ctx)

# pull out and inspect the state
state = await ctx.get("state")
print(state["num_fn_calls"])
```
````
